### PR TITLE
Deprecate -join and -join-wan

### PR DESCRIPTION
**Pull request by pglass**
### Description

This deprecates the `-join` and `-join-wan` flags and the corresponding `start_join` and `start_join_wan` config fields. Instead, `-retry-join` or `-retry-join-wan` should be used, or the corresponding `retry_join` or `retry_join_wan` config fields.

This is a breaking change. The deprecated options are now aliases of the "retry join" variants:

* `-join` is now an alias of `-retry-join`
* `-join-wan` is now an alias of `-retry-join-wan`
* `start_join` is now an alias of `retry_join`
* `start_join_wan` is now an alias of `retry_join_wan`

When the deprecated flags/fields are used:

* Warning(s) are logged.
* LAN addresses from `-join` and `start_join` are merged with the list of addresses specified by `-retry-join` and `retry_join`
* WAN addresses from `-join-wan` and `start_join_wan` are merged with the list of addresses specified by `-retry-join-wan` and `retry_join_wan`

### PR Checklist

* [x] updated test coverage
* [x] external facing docs updated
* [x] not a security concern

### CODE REVIEW

File: agent/auto-config/auto_encrypt.go [hunk 1]
```diff
@@ -88,7 +88,7 @@ func (ac *AutoConfig) joinHosts() ([]string, error) {
 	// addresses. These are for joining serf and therefore we cannot rely on the
```
Comment (rboyer) [inline agent/auto-config/auto_encrypt.go:88]: The line above this one references "start join" in a comment.

File: website/content/docs/agent/config/config-files.mdx [hunk 2]
```diff
@@ -2220,3 +2218,7 @@ server.
     The default value is `250`. It is based on a load test of 5,000 streams connected to a single server with two CPU cores.
 
     If necessary, you can lower or increase the limit without a rolling restart by using the `consul reload` command or by sending the server a `SIGHUP`.
+
+<!-- list of reference-style links -->
```
Comment (rboyer) [inline website/content/docs/agent/config/config-files.mdx:2222]: Why is this included?

Comment (pglass) [reply website/content/docs/agent/config/config-files.mdx:2222]: Oh, this fixes broken links I noticed. There are a bunch of places on this page where `[go-sockaddr]` is being rendered literally in the HTML:

Screen Shot 2022-12-07 at 4 42 18 PM

Although, I should make this doc fix a separate PR and backport it. I'll do that instead

Comment (pglass) [reply website/content/docs/agent/config/config-files.mdx]: Ok, moved this fix to https://github.com/hashicorp/consul/pull/15723

**Review by rboyer — APPROVED** [review]

File: website/content/docs/enterprise/license/overview.mdx [hunk 3]
```diff
@@ -48,18 +48,18 @@ Virtual agents do not need the license to run.
...
 
 #### Client Agent License Retrieval
 
 When a client agent starts without a license in its configuration or environment, it will try to retrieve the
 license from the servers via RPCs. That RPC always requires a valid non-anonymous ACL token to authorize the
 request but the token doesn't need any particular permissions. As the license is required before the client
-actually joins the cluster, where to make those RPC requests to is inferred from the [`start_join`](/docs/agent/config/config-files#start_join)
-or [`retry_join`](/docs/agent/config/config-files#retry_join) configurations. If those are both unset or no
+actually joins the cluster, where to make those RPC requests to is inferred from the
+[`retry_join`](/docs/agent/config/config-files#retry_join) configurations. If those are both unset or no
```
Comment (kisunji) [inline website/content/docs/enterprise/license/overview.mdx:62]:
```suggestion
[`retry_join`](/docs/agent/config/config-files#retry_join) configuration. If `retry_join` is unset or no
```

**Review by kisunji — APPROVED** [review]
minor nits but LGTM

**Review by skpratt — APPROVED** [review]
